File: Databases/extremedb/eXtremeDB/samples/python/open/open.py
import os
import sys
import exdb

#define eXtremeDB schema - exactly like in .mco file
schema = '''
#define uint4     unsigned<4>

declare database opendb;

class Class1  {
    uint4 i4;
};
'''

is_disk = False
tmgr = 'mursiw'
is_shm = False
is_debug = False

#Load Runtime configuration specified by parameters
exdb.init_runtime(is_disk, tmgr, is_shm, debug=is_debug)

# Parse schema, load dictionary and create python dictionary object
dict = exdb.load_dictionary(schema, persistent=is_disk, debug=is_debug)

print 'List structures in the loaded dictionary'
for s in dict.structs:
    print 'type ', type(dict.structs[s])
    print "%s=>%s - %s" % (s, dict.structs[s], dict.structs[s].__dict__)
    for fld in dict.structs[s].__dict__:
        f = dict.structs[s].__dict__[fld]
        if issubclass(type(f), exdb.eXtremeDBField):
            print "fld %s=>%s" % (fld, f.__dict__)

print 'List classes in the loaded dictionary'
for s in dict.classes:
    cls = dict.classes[s]
    print "%s=>%s - %s" % (s, cls, cls.struct.__dict__)
    for fld in cls.struct.__dict__:
        f = cls.struct.__dict__[fld]
        if issubclass(type(f), exdb.eXtremeDBField):
            print "fld %s=>%s" % (fld, f.__dict__)

# Open (create) database with default parameters

db = exdb.open_database(dbname='opendb', dictionary=dict, is_disk=is_disk, db_segment_size=128*1024*1024)
    
print 'Database created'
#Connect to the database 
conn = None
with db.connect() as conn:

    print 'Database connected. Connection object is', conn

# Open transaction. By default transaction is opened in Read-Write mode
    conn.startTransaction()

# Create an object of type specified in the schema definition
    o = conn.new("Class1")

# Print field(s) of created object with fields filled with default values
    print o   
    
# Fields in object are filled with default values, so we can now save object in database
    conn.commit() 

# Connection will automatically close when leaving scope of with statement.
# It may be handled manually if use:
# conn = db.connect()
# ....
# conn.close()    

print 'Closing database'
db.close()
print 'Done'

File: Databases/extremedb/eXtremeDB/samples/lua/open/open.lua
require ('exdb')

is_debug = false
is_disk = false
is_shm = false

tmgr = 'mursiw'

function rmfile( fname )
    local f = io.open(fname, "r")
    if f then
        f:close()
        os.remove(fname)
        print("File deleted:", fname)
    end
end

rmfile('opendb.dbs')
rmfile('opendb.log')

print ("debug=", is_debug)
initRuntime{disk=is_disk,debug=is_debug,shm=is_shm,tmgr=tmgr}

print ("exdb initialized")
-- devices are a table list hence {{ ...}}. Possible options see in exdb.lua
devices = {{  type='conv',
              assignment= 'database', 
              size = 64*1024*1024  
           }}
-- db params is also a table
dbparams = {ddl_dict_size=64*1024, max_classes=100, max_indexes=100}

-- for persistent databases additinal devices are necessary
if is_disk then
    table.insert(devices, {type='conv', assignment='cache', size=64*1024*1024})
    table.insert(devices, {type='file', assignment='persistent', name='opendb.dbs'})
    table.insert(devices, {type='file', assignment='log', name='opendb.log'})

-- and addition paparemets
    dbparams.disk_page_size = 4096
end
-- finally create a database
db = create("luadb", devices, dbparams)


print ("database created", db, devices[1].ptr)

-- connect to it
c = connect{db=db}

--  c is a database connection and now can be used to perform SQL statements
print (c)

print ("Creating table")

c:statement("create table t(i int)")

print ("Inserting value")
c:statement("insert into t values([1,2,3])")

print ("Selecting count, db = ", db)

-- quering, cursor is a resut of thequery
curs = c:query("select count(*) as nrecs from t")

--print ("Select performed, curs = ", curs, "count=", curs:next().nrecs)
    
--
-- another query
--
print ("Selecting, db = ", db)
curs = c:query("select i from t")
print ("Select performed, curs = ", curs)
for elem in curs:rows() do
    print ("i = ", elem.i)
end

print ("Done. Closing database")
c:close()

close(db)
close('luadb')

print ("Free resources")
-- Free database memory with LUA GC
devices[1].ptr = nil


print ("Done")
